py/dml/expr.py

```diff
@@ -228,6 +229,74 @@ def typecheck_inargs(site, args, inp, kind="function", known_arglen=None):
         else:
             raise EPTYPE(site, arg, rtype, pname, kind)
 
+def typecheck_inarg_inits(site, inits, inp, location, scope,
+                          kind="function", variadic=False,
+                          allow_undefined_args=False):
+    if (not variadic and len(inits) != len(inp)) or len(inits) < len(inp):
+        raise EARG(site, kind)
+
+    from .expr_util import coerce_if_eint
+    from .codegen import eval_initializer, codegen_expression, \
+                         codegen_expression_maybe_nonvalue
+
+    args = []
+    for (init, (pname, ptype)) in zip(inits, inp):
+        if ptype is None:
+            if init.kind != 'initializer_scalar':
+                raise ESYNTAX(init.site, '{',
+                              'the argument for an untyped parameter must be '
+                              + 'a simple expression')
+            arg = codegen_expression_maybe_nonvalue(init.args[0], location,
+                                                    scope)
+            if (isinstance(arg, NonValue)
+                and not (allow_undefined_args and arg.undefined)):
+                raise arg.exc()
+            args.append(arg)
+        elif (not dml.globals.compat_dml12_int(site)
+            or not dml.globals.compat_dml12_int(init.site)
+            or init.kind != 'initializer_scalar'):
```

Comment:
perhaps `X and Y and Z==W` is easier to read than `not X or not Y or Z!=W` (if you flip true/false branch)

Reply:
Agreed.